### ISSUE TITLE
Before further work

### ISSUE THREAD

**Issue by antitoxic** [developit/preact-router#2]
A common pattern I've noticed around JSX routers is to use `<Router>`, `<Route>` , `<Link>` .

I wanted to ask: Isn't `<Link>` enough? 

Unmounting and mounting entire page components via `path=""` or `<Route>` has 2 drawbacks:

1. What if my `<About>` page and my `<TermsOfService>` page only differ by `<Title>` and `<PageText>`? Why do I swap the entire page? Especially if my `<Title>` and `<PageText>` are not sibling elements and have different parents. One is in the header, the other is in the main body section. Both pages migh also share other common elements like in-page sidebar. Additionally swapping entire page removes the ability to only animate the changed elements, not the entire page.
2. Routing casues a DOM mutation that  is triggered not by store/redux state change (via action) but custom logic (`<Route>`). 

Is it crazy to propose just a single element `<Link>` that has a `onPageNavigation` contextType  property which in decide what to do when page changes. If someone uses redux, they will set `onPageNavigation` to execute a `PAGE_NAVIGATION_STARTED` action, possibly supplying route name if named.

**Comment by developit** [comment]
Interesting, and I see your point. Your suggestion seems more analogous to a state-based router than a URL-based router. Most of my previous work with other frameworks has centered around a more abstract state where the URL is simply treated as a source of changes, similar to what you described.

My intention with `preact-router` was to create a minimal URL router, but I would be happy to work on an alternative with semantics similar to what you outlined - perhaps a second project would suit this?

**Comment by antitoxic** [comment]
I think the whole of React and all general component-based apps base their rendering on data (*commonly called state*). URL is nothing more that new data.

**What does changing the URL means?** - It means data changes, like changing `Title` and `Content` changes. The concept of *"page"* is fake. There is no such thing as page. In the days before SPA (*ironically page is in the name*) in order to show new data you need to reload the entire html. That's where the misconception that changing the urls means changing the entire root element.

Another proof is that good APIs don't give you a `?page=..` parameter. They give you `?start=..&limit=...`. No page.

If my URL is `/posts/show/new` it could be that I show the same thing as on `/posts/` but I show a form for adding a new post in a modal dialog. 

`preact-router` and other routes shouldn't promote page-based behavior. It's an antipattern.

**Comment by developit** [comment]
While I agree that it's great to avoid special-casing the URL, that is not an approach I would ever force people to take.  It is also absolutely not an antipattern, it's simply a different way of composing UI.

Your point about a page parameter seems to have nothing to do with this discussion at all, besides conflating the concepts of "paging" and "navigation". Plus, well-designed APIs do have semantic URLs, and they are given special meaning (resource type) - that is the entire purpose of REST.

One very important thing I want to clarify: you are not correct regarding "changing the entire root element" in response to a page. Re-rendering from the root is intentionally how composition works in Preact, or even React for that matter - that is the entire reason for using Virtual DOM diffing.  Here's a very clear visual demonstration that changing the URL absolutely still performs the same diff as any update you might observe from a less maligned data source like redux:

![example](https://i.gyazo.com/9429fd715b2c66f7f3caf97ac7b3a585.gif)

Lastly, on the topic of promotion: this is a library that allows a number of patterns for responding to URL changes. It's likely the README does a disservice in showing newcomers what is surely the easiest way to deal with URLs - loading components in response to them automatically.  However, the library is really just a way to observe changes in the URL and invoke a function.  Here are a few examples of other ways to use `preact-router` that edge more towards your suggestions:

```js
export default new Router({
    onChange({ url, params }) {
        switch(url) {
            case "/something": action({ params });
            // etc
        }
    }
});
```

or, if you'd like to still use JSX and define your routes while not having them prescribe composition:

```js
const Route = ({ action, ...params }) => {
    actions[action](params);
    return null;
};

(
    <Router>
        <Route path="/" action="home" />
        <Route path="/profile" action="showProfiles" />
        <Route path="/profile/:user" action="showProfile" />
        <Route path="/login" action="showLogin" />
    </Router>
);
```

Hope this helps clear up the difference between the documented "easy" usage and the many other possible ways to use this router.

**Comment by antitoxic** [comment]
> Hope this helps clear up the difference between the documented "easy" usage and the many other possible ways to use this router.

@developit It defenitely does. Thank you. I'm sorry if I sounded pushy. Just wanted to emphasize on the a problem that I saw. 

I didn't know that if I swap `<AboutPage>` with `<TOCPage>`, preact would detect that only title has changed and replace only that? Is that true? I mean, wouldn't the vdom detect that `TOCPage` is not `AboutPage` and swap the first with the latter, not only the titles in the first? By `root element` I meant the `pageElement` which is usually the main area of an app.  I'm guess in the gif above the page element is just a section of the app.

PS: `page` parameter has at least some relevance in the discussion. It shows that in software development "paging" has no meaning, i.e. resources have no pages,  they are simply entire resources or parts of them limited by parameters. I'm well acquainted with REST and in fact I started the discussion due to that. It goes hand in hand with stateless behavior, one that only rely on data received and passed along. I understand the purpose of the router is to also help newcomers, and you gave compelling examples of how advanced scenarios can be achieved.
